infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java

```diff
@@ -0,0 +1,127 @@
+
+    pods.values().forEach(e -> podData.add(new PodData(e.getSpec(), e.getMetadata())));
+
+    deployments
+        .values()
+        .stream()
+        .map(d -> d.getSpec().getTemplate())
+        .forEach(pt -> podData.add(new PodData(pt.getSpec(), pt.getMetadata())));
+
+    return merge(createDeployment(), podData);
+  }
+
+  public Deployment merge(Deployment baseDeployment, List<PodData> pods) {
+    PodTemplateSpec basePodTemplate = baseDeployment.getSpec().getTemplate();
+    ObjectMeta baseMeta = basePodTemplate.getMetadata();
+    PodSpec baseSpec = basePodTemplate.getSpec();
+
+    Set<String> containerNames =
+        baseSpec
+            .getContainers()
+            .stream()
+            .map(Container::getName)
+            .collect(toCollection(HashSet::new));
+    Set<String> initContainerNames =
+        baseSpec
+            .getInitContainers()
+            .stream()
+            .map(Container::getName)
+            .collect(toCollection(HashSet::new));
+    for (PodData podData : pods) {
+      baseMeta.getLabels().putAll(podData.getMetadata().getLabels());
+      baseMeta.getAnnotations().putAll(podData.getMetadata().getAnnotations());
+      baseMeta.getAdditionalProperties().putAll(podData.getMetadata().getAdditionalProperties());
+
+      for (Container container : podData.getSpec().getContainers()) {
+        // store original recipe machine name
+        baseMeta
+            .getAnnotations()
+            .put(
+                format(MACHINE_NAME_ANNOTATION_FMT, container.getName()),
+                Names.machineName(podData.getMetadata(), container));
+
+        //generate container name to avoid collisions
+        while (!containerNames.add(container.getName())) {
+          container.setName(NameGenerator.generate(container.getName(), 4));
+        }
+        baseSpec.getContainers().add(container);
+      }
+
+      for (Container initContainer : podData.getSpec().getInitContainers()) {
+        //generate container name to avoid collisions
+        while (!initContainerNames.add(initContainer.getName())) {
+          initContainer.setName(NameGenerator.generate(initContainer.getName(), 4));
+        }
+        baseSpec.getInitContainers().add(initContainer);
+      }
+
+      baseSpec.getVolumes().addAll(podData.getSpec().getVolumes());
+      baseSpec.getImagePullSecrets().addAll(podData.getSpec().getImagePullSecrets());
+
+      // ??
+      baseSpec.getTolerations().addAll(podData.getSpec().getTolerations());
+      baseSpec.getNodeSelector().putAll(podData.getSpec().getNodeSelector());
+      baseSpec.getAdditionalProperties().putAll(podData.getSpec().getAdditionalProperties());
+    }
+
+    baseDeployment.getSpec().getSelector().setMatchLabels(baseMeta.getLabels());
+
+    return baseDeployment;
+  }
+
+  private Deployment createDeployment() {
+    return new DeploymentBuilder()
+        .withMetadata(new ObjectMetaBuilder().withName("workspace").build())
```

Comment:
Name collisions for deployments of multple workspaces?

Reply:
Each deployment will be renamed with generated name before a start https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/UniqueNamesProvisioner.java#L90